lib/train/transports/ssh_connection.rb

```diff
@@ -334,7 +336,20 @@ def execute_on_channel(cmd)
           end
         end
       end
-      session.loop
+
+      thr = Thread.new { session.loop }
+      
```

Comment:
I ran some of this in a few test files and after a think I do believe this is a proper and safe use of Ruby's `Thread`.

Much of the threadsafety issues I would worry about have to do with asynchronous interruptions, creating state (especially if the state is passed around) and managing multiple threads. 

Since this new code keeps it to the one thread with no fancy state moving in/out/around the thread and its contents, that should be good. 

As I understand it `#join` pauses any further code execution until the contents of the thread block either finish or the limit (timeout) is reached. In that case, I believe we wouldn't have to worry about other Threads getting jammed on the back of this one. 

I am happy to swap this PR to an approval any time but will leave it currently as a comment as I know we've requested a review from Foundations to ensure this plays well with the rest of the ecosystem. :)

Reply:
This PR is almost entirely different from when you first looked at it; so taking another look before approving would be appreciated, thanks!

As we were seeing earlier, this change needs to work well for a. a subsequent change to `train-winrm` and b. all our current plugin users